bridge-token-factory/src/fee.rs

```diff
@@ -0,0 +1,661 @@
+                    lower_bound,
+                    upper_bound,
+                },
+            },
+        );
+    }
+
+    // Fee should be added as per: 10% -> 0.1 = 0.1*10^6 with proper fee amount bounds
+    #[access_control_any(roles(Role::FeeSetter))]
+    #[payable]
+    pub fn set_withdraw_fee(
+        &mut self,
+        token: EthAddressHex,
+        fee_percentage: U128,
+        lower_bound: Option<U128>,
+        upper_bound: Option<U128>,
+    ) -> Promise {
+        self.withdraw_fee.insert(
+            &token.0,
+            &Fee {
+                fee_percentage,
+                bounds: FeeBounds {
+                    lower_bound,
+                    upper_bound,
+                },
+            },
+        );
+
+        require!(env::attached_deposit() >= self.bridge_token_storage_deposit_required);
+        ext_bridge_token::ext(self.get_bridge_token_account_id(&token))
+            .with_static_gas(TOKEN_STORAGE_DEPOSIT_GAS)
+            .with_attached_deposit(env::attached_deposit())
+            .storage_deposit(None, None)
+    }
+
+    // Fee should be added as per: 10% -> 0.1 = 0.1*10^6 with proper fee amount bounds
+    #[access_control_any(roles(Role::FeeSetter))]
+    pub fn set_withdraw_fee_per_silo(
+        &mut self,
+        silo_account_id: AccountId,
+        token: Option<EthAddressHex>,
+        fee_percentage: U128,
+        lower_bound: Option<U128>,
+        upper_bound: Option<U128>,
+    ) {
+        self.withdraw_fee_per_silo.insert(
+            &get_silo_fee_map_key(&silo_account_id, token.as_ref()),
+            &Fee {
+                fee_percentage,
+                bounds: FeeBounds {
+                    lower_bound,
+                    upper_bound,
+                },
+            },
+        );
+    }
+
+    pub fn get_deposit_token_fee(&self, token: &EthAddressHex) -> Option<Fee> {
+        self.deposit_fee.get(&token.0)
+    }
+
+    pub fn get_withdraw_token_fee(&self, token: &EthAddressHex) -> Option<Fee> {
+        self.withdraw_fee.get(&token.0)
+    }
+
+    // Returns withdraw fee for different tokens per silo
+    pub fn get_withdraw_fee_per_silo(
+        &self,
+        silo_account_id: AccountId,
+        token: Option<EthAddressHex>,
+    ) -> Option<Fee> {
+        self.get_withdraw_fee_per_silo_internal(&silo_account_id, token.as_ref())
+    }
+
+    // Returns desposit fee for different tokens per silo
+    pub fn get_desposit_fee_per_silo(
+        &self,
+        silo_account_id: AccountId,
+        token: Option<EthAddressHex>,
+    ) -> Option<Fee> {
+        self.get_deposit_fee_per_silo_internal(&silo_account_id, token.as_ref())
+    }
+
+    // Accumulated fee should be claimed from here.
+    #[access_control_any(roles(Role::FeeClaimer))]
+    pub fn claim_fee(&self, token: AccountId, amount: Balance) {
+        ext_bridge_token::ext(token)
+            .with_static_gas(FT_TRANSFER_GAS)
+            .with_attached_deposit(1)
```

Comment:
Who provided this  One YoctoNEAR? Should this function be payable?

Reply:
5c69b52e3017e6bcb8c0bf27f65f53065eed5158